src/main/scala/com/databricks/spark/xml/util/TypeCast.scala

```diff
@@ -116,11 +116,32 @@ private[xml] object TypeCast {
   )
 
   private def parseXmlTimestamp(value: String, options: XmlOptions): Timestamp = {
-    val formatters = options.timestampFormat.map(DateTimeFormatter.ofPattern).
-      map(supportedXmlTimestampFormatters :+ _).getOrElse(supportedXmlTimestampFormatters)
-    formatters.foreach { format =>
+    // Loop over built-in formats
+    supportedXmlTimestampFormatters.foreach { format =>
+      try {
+        return Timestamp.from(
+          ZonedDateTime.parse(value, format).toInstant
+        )
+      } catch {
+        case _: Exception => // continue
+      }
+    }
+    // Custom format
+    if (options.timestampFormat.isDefined) {
+      val format = DateTimeFormatter.ofPattern(options.timestampFormat.get)
+      try {
+        // Custom format with timezone or offset
+        return Timestamp.from(
+          ZonedDateTime.parse(value, format).toInstant
+        )
+      } catch {
+        case _: Exception => // continue
+      }
       try {
-        return Timestamp.from(ZonedDateTime.parse(value, format).toInstant)
+        // Custom format without timezone or offset
+        return Timestamp.from(
+          ZonedDateTime.parse(value, format.withZone(ZoneId.of(options.timezone.get))).toInstant
+        )
```

Comment:
We can check if there is a timezone but there is no method to determine if an offset is defined

Reply:
Hm. What if we parse, and if it fails with the error you are facing, add 'withZone'? the only problem is that may be a huge amount of perf overhead, so we'd have to have some way to test and store the format with the zone, only if it's needed. Not pretty but that would be the way forward

Reply:
When parsing fails, it will always throw a `DateTimeParseException`, no matter if it is because of the problem we are talking about or not. Rather than parsing the error message, I wrote a `isParseableAsZonedDateTime()` method and used it during the loop.